### ISSUE TITLE
Implement AnglE loss

### ISSUE THREAD

**Issue by tomaarsen** [UKPLab/sentence-transformers#2437]
Hello!

## Context
This issue describes a feature that I am planning to be included in a release before v3, or alternatively, in v3 of Sentence Transformers.

## Details
Some [recent work by Li & Li, 2023](https://arxiv.org/abs/2309.12871) introduces an AnglE objective. Its inclusion in Sentence Transformers would be beneficial to allow our users to also train using this new loss function. The first author, @SeanLee97, has implemented the objective in https://github.com/SeanLee97/AnglE.

@johneckberg has expressed an interest in contributing this loss.

cc @bwanglzu @ir2718 @johneckberg @aamir-s18 as I know you're interested in my TODO list. John will lead this development, and I will review his PR. Additionally, @SeanLee97 may also have a look if he finds the time.

- Tom Aarsen

**Comment by johneckberg** [comment]
I will get this out for review in the coming days!

AnglE loss is CoSENT loss with a novel pairwise similarity metric. In the proposed code for CoSENT loss I allow for the choice of similarity metric in the constructor. In my head, it might make sense to implement the AnglE metric calculation in utils, where a user can specify it as a metric instead of the default pairwise_cos_sim when using CoSENT loss. What do others think of that?

**Comment by tomaarsen** [comment]
If that similarity metric has a similar signature as e.g. [cosine similarity](https://github.com/UKPLab/sentence-transformers/blob/b00b329ffb27e3f48d841781224a9676deed5833/sentence_transformers/util.py#L30-L51), then it may make sense to discuss this with @ir2718 who is planning on creating a new Enum with all similarity measures in #2441.

I would be in favor of this approach.

We can then also introduce an AnglELoss class which is just a subclass of CoSENT, but with the similarity measure fixed to the pairwise similarity metric. How does that sound?

- Tom Aarsen

**Comment by johneckberg** [comment]
Sounds great Tom! I'll get started on that soon

**Comment by dawnik17** [comment]
The angle difference in util.pairwise_angle_sim is implemented as follows, 

`loss = logsum(1 + exp(s(k,l) - s(i,j)))`, where (i,j) and (k,l) are any of the input pairs in the batch such that the expected similarity of (i,j) is greater than (k,l).

But in the paper, it actually is represented as `logsum(1 + exp(s(i,j) - s(k,l)))`

I'm attaching a screenshot from the paper as well. I might be understanding it incorrectly, and would appreciate some clarification. Thanks in advance :) @tomaarsen @johneckberg 

![Screenshot 2024-08-22 at 6 26 45 PM](https://github.com/user-attachments/assets/2a92d5f3-244c-4505-badc-02c3a2337449)

**Comment by tomaarsen** [comment]
Hello @dawnik17. Apologies for the delay.
If you remember, are you proposing that https://github.com/UKPLab/sentence-transformers/blob/2e13ee61d92ea9b5bcd33887aaece3af263dc85c/sentence_transformers/losses/CoSENTLoss.py#L85
should be 
```python
scores = scores[None, :] - scores[:, None]
```
to match the paper?

If so, the current implementation does give notably better performance when training a simple model.

- Tom Aarsen

**Comment by johneckberg** [comment]
@dawnik17 I am so sorry I missed this! You are correct, in the paper, it's `logsum(1 + exp(s(i,j) - s(k,l)))`. This is a documentation error on my part in the `CoSENTLoss.py` file.  The doc should read `loss = logsum(1+exp(s(i,j)-s(k,l))+exp...)` not `loss = logsum(1+exp(s(k,l)-s(i,j))+exp...)`.
